hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -266,8 +313,12 @@ public final void writeObject(final ObjectDataOutput out, final Object obj) {
         if (obj instanceof Data) {
             throw new HazelcastSerializationException("Cannot write a Data instance, use writeData() instead");
         }
+        // We don't expect obj being compact serializable is a common usage.
+        // (ObjectDataOutput.writeObject(somethingCompactSerializable) ).
+        // If obj is compact serializable, we include the schema always to be on the safe side.
+        // Otherwise, if this API is used and the data is stored on HotRestart or send via WAN, we couldn't reach the schema.
+        SerializerAdapter serializer = serializerFor(obj, true);
         try {
-            SerializerAdapter serializer = serializerFor(obj);
```

Comment:
To avoid unnecessary catch and rethrow. If am not missing the point, anything serializerFor throws is re-thrown as is in the `handleSerializeException` anyway.

Reply:
If we were to throw `new HazelcastSerializationException("There is no suitable serializer for " + type);`, it was wrapped in another exception with a different message. I think it is OK to change the error message, so this should not be a problem.